Xkcd-Comics/ContentViews/ComicsView.swift

```diff
@@ -0,0 +1,127 @@
+//
+//  ContentView.swift
+//  Xkcd-Comics
+//
+//  Created by Jia-Jiuan Tsai on 2021-11-24.
+//
+
+import SwiftUI
+import Kingfisher
+
+struct ComicsView: View {
+    @ObservedObject var viewModel: ComicsViewModel
+    
+    var body: some View {
+        NavigationView {
+            VStack {
+                Spacer()
+                ComicView(viewModel: viewModel)
+                Spacer()
+                NextAndPreviousButtonView(viewModel: viewModel)
+            }.navigationTitle(viewModel.comic.title)
+            .onTapGesture {
+                viewModel.toggleInfo()
+            }
+            .navigationBarItems(trailing: NavigationBarView(viewModel: viewModel))
+        }
+    }    
+}
+
+struct ComicsView_Previews: PreviewProvider {
+    static var previews: some View {
+        ComicsView(viewModel: ComicsViewModel(
+                    fetchingService: FetchingService(urlSession: URLSession.shared,
+                                                     jsonDecoder: JSONDecoder()),
+                    userDataService: UserDataService(userDefaults: UserDefaults.standard,
+                                                     jsonEncoder: JSONEncoder(),
+                                                     jsonDecoder: JSONDecoder())))
+    }
+}
+
+// MARK: Extracted Subviews
+struct ComicView: View {
+    @ObservedObject var viewModel: ComicsViewModel
+    
+    var body: some View {
+        if viewModel.isShowingInfo {
+            VStack {
+                Text(viewModel.comic.alt)
+                    .font(.title2)
+                    .bold()
+                    .padding()
+                
+                Text("Comic Number: \(viewModel.comic.num) Created: \(viewModel.comic.month )/\(viewModel.comic.year)")
+                    .foregroundColor(.gray)
+                
+                HStack {
+                    CustomButton(icon: "info.circle", title: "Explanation") {
+                        viewModel.toggleDescription()
+                    }.sheet(isPresented: $viewModel.isShowingExplanation) {
+                        ExplanationView(viewModel: ExplanationViewModel(number: viewModel.comic.num))
+                    }
+                }
+            }
+        } else {
+            KFImage(URL(string: viewModel.comic.img))
+                .resizable()
+                .scaledToFit()
+                .padding()
+        }
+    }
+}
+
+struct NextAndPreviousButtonView: View {
```

Comment:
This view is declared as a stand-alone type that is possible to instantiate anywhere in the app - not just inside ComicsView. Its design is also very generic - two buttons with a design for going backwards and forwards in some feed. However, as the code is written right now, it is very tightly coupled to how ComicsView works, and requires an entire ComicsViewModel, and so it would not be usable for any other similar views. 

If all you want to do is divide the code up into smaller parts, and not expose this code stand alone to the rest of the app, perhaps it would be more suitable to make this a private struct inside ComicsView, or a private method.

```swift
@ViewBuilder private func nextAndPreviousButtons() -> some View {
    HStack {
    ...
   }
}
```

On the other hand, I do think this view could be refactored into a more generic, re-usable view. Instead of taking the entire ComicsViewModel as a parameter, it could take two function parameters

```swift
let previousButtonSelected: () -> Void
let nextButtonSelected: () -> Void
```

It would then be the responsibility of `ComicsView` to set up this view and provide two closures that define what the buttons should do in that context, and other views could provide other closures. 

It would also be the responsibility of `ComicsView`/ `ComicsViewModel` to decide what to do when fetching comics fail - display alerts, etc, without the buttons having to know about that. 

Typically, the smaller a view gets, the dumber it should be. A button group widget should just be two dumb buttons that don't know anything about any business logic, but are just buttons that report back when they have been clicked.